src/entry_stream.rs

```diff
@@ -3,58 +3,421 @@
 //! real-time access to entries.
 
 use crate::entry::Entry;
+use crate::leader_scheduler::LeaderScheduler;
+use crate::leader_scheduler::DEFAULT_TICKS_PER_SLOT;
 use crate::result::Result;
-use chrono::Utc;
+use chrono::{SecondsFormat, Utc};
+use solana_sdk::hash::Hash;
+use solana_sdk::pubkey::Pubkey;
 use std::io::prelude::*;
 use std::net::Shutdown;
 use std::os::unix::net::UnixStream;
 use std::path::Path;
 
 pub trait EntryStreamHandler {
-    fn stream_entries(&mut self, entries: &[Entry]) -> Result<()>;
+    fn emit_entry_events(
+        &mut self,
+        leader_scheduler: &LeaderScheduler,
+        entries: &[Entry],
+    ) -> Result<()>;
+    fn emit_slot_event(
+        &mut self,
+        current_slot: u64,
+        tick_height: u64,
+        leader_id: &Pubkey,
+    ) -> Result<()>;
+    fn emit_block_event(&mut self) -> Result<()>;
+}
+
+trait Output {
+    fn write(&mut self, payload: String) -> Result<()>;
+}
+
+trait EntryStreamEvents {
+    fn do_entry_events(
+        &mut self,
+        output: &mut Output,
+        leader_scheduler: &LeaderScheduler,
+        entries: &[Entry],
+    ) -> Result<()>;
+    fn do_slot_event(
+        &mut self,
+        output: &mut Output,
+        current_slot: u64,
+        tick_height: u64,
+        leader_id: &Pubkey,
+    ) -> Result<()>;
+    fn do_block_event(&mut self, output: &mut Output) -> Result<()>;
+}
+
+impl EntryStreamEvents for EntryStreamMeta {
+    fn do_entry_events(
+        &mut self,
+        output: &mut Output,
+        leader_scheduler: &LeaderScheduler,
+        entries: &[Entry],
+    ) -> Result<()> {
+        for entry in entries {
+            let (_prev_leader_id, prev_slot) = leader_scheduler
+                .get_scheduled_leader(self.last_tick_height)
+                .unwrap();
+            let (curr_leader_id, curr_slot) = leader_scheduler
+                .get_scheduled_leader(entry.tick_height)
+                .unwrap();
```

Comment:
@CriesofCarrots @mvines this is where I'm wondering why the slots are not returned properly; any thoughts?

Reply:
This method actually doesn't exist anymore on the tip of master, so I'd hate to see you spend anymore of your time troubleshooting it!
I'm working on incorporating @mvines' patches and rebasing your changes. Should have something helpful for you soon.

Reply:
@sunnygleason , I have some code that works with real nodes on the tip of master. Working on updating your test now.